timestamp-api/src/main/java/com/palantir/timestamp/AbstractTimestampBoundStoreWithId.java

```diff
@@ -0,0 +1,203 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.timestamp;
+
+import java.util.UUID;
+import java.util.function.Function;
+
+import javax.annotation.Nullable;
+
+import org.immutables.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.helpers.MessageFormatter;
+
+import com.palantir.util.debug.ThreadDumps;
+
+public abstract class AbstractTimestampBoundStoreWithId implements TimestampBoundStore {
+    protected long currentLimit = -1;
+    protected boolean startingUp = true;
+    protected UUID id;
+    protected long initialValue;
+
+    public static Logger log = LoggerFactory.getLogger(TimestampBoundStore.class);
+
+    protected AbstractTimestampBoundStoreWithId(Long initialValue) {
+        this.id = UUID.randomUUID();
+        this.initialValue = initialValue;
+    }
+
+    public UUID getId() {
+        return id;
+    }
+
+    @Override
+    public synchronized long getUpperLimit() {
+        log.debug("[GET] Getting upper limit");
+        return runReturnLong(client -> {
+                    TimestampBoundStoreEntry entryInDb = getStoredEntry(client);
+                    entryInDb = migrateIfStartingUp(entryInDb);
+                    checkMatchingId(entryInDb);
+                    setCurrentLimit("GET", entryInDb);
+                    return currentLimit;
+                }
+        );
+    }
+
+    @Override
+    public synchronized void storeUpperLimit(final long limit) {
+        log.debug("[PUT] Storing upper limit of {}.", limit);
+        checkAndSet(TimestampBoundStoreEntry.create(currentLimit, id),
+                TimestampBoundStoreEntry.create(limit, id));
+    }
+
+    private TimestampBoundStoreEntry migrateIfStartingUp(TimestampBoundStoreEntry entryInDb) {
+        if (!startingUp) {
+            return entryInDb;
+        }
+        log.info("[GET] The service is starting up. Attempting to get timestamp bound from the DB and"
+                + " resetting it with this process's ID.");
+        TimestampBoundStoreEntry newEntry = TimestampBoundStoreEntry.create(
+                entryInDb.getTimestampOrValue(initialValue), id);
+        checkAndSet(entryInDb, newEntry);
+        startingUp = false;
+        return newEntry;
+    }
+
+    private void checkAndSet(TimestampBoundStoreEntry entryInDb, TimestampBoundStoreEntry newEntry) {
+        cas(entryInDb, newEntry);
+    }
+
+    private void cas(TimestampBoundStoreEntry entryInDb, TimestampBoundStoreEntry newEntry) {
+        checkLimitNotDecreasing(newEntry);
+        log.info("[CAS] Trying to set upper limit from {} to {}.", entryInDb.getTimestampAsString(),
+                newEntry.getTimestampAsString());
+        Result result = updateEntryInDb(entryInDb, newEntry);
```

Comment:
Looks like this could return null in C* impl, which would NPE. Should have tests with coverage over these paths.